apis/Google.Cloud.Storage.V1/Google.Cloud.Storage.V1/UrlSigner.PostPolicy.cs

```diff
@@ -0,0 +1,315 @@
+﻿// Copyright 2020 Google LLC
+// 
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+// 
+//     https://www.apache.org/licenses/LICENSE-2.0
+// 
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+using Google.Api.Gax;
+using Newtonsoft.Json;
+using System;
+using System.Collections.Generic;
+using System.Collections.ObjectModel;
+using System.Linq;
+using System.Net;
+
+namespace Google.Cloud.Storage.V1
+{
+    public sealed partial class UrlSigner
+    {
+        /// <summary>
+        /// A known element that can be added to a <see cref="PostPolicy"/>.
+        /// </summary>
+        public enum PostPolicyKnownElement
+        {
+            /// <summary>
+            /// The ACL that you want to allow for the objects that will be uploaded.
+            /// </summary>
+            Acl = 0,
+
+            /// <summary>
+            /// The name of the bucket that you want to allow uploads to.
+            /// Alternatively, you can include the bucket name when you add 
+            /// the <see cref="PostPolicyKnownElement.Key"/> element.
+            /// </summary>
+            BucketName = 1,
```

Comment:
Thanks for clarifying. I'm pending the surface examples to parse the design better, but in other languages Bucket is a required parameter as is Key:

```python
client.generate_signed_post_policy_v4(
    bucket_name="rsaposttest-1579902678-pt5yms55j47r6qy4",
    blob_name="test-object",
    expiration=datetime.datetime.now() + datetime.timedelta(seconds=10),
    fields={"success_action_status": 200},
)
```

Reply:
Ah but I believe the bucket cannot be required because it can be specified as part of the key. In that case it goes into the policy as part of the key so everything is still secured, but there's no need to specify the bucket separately.

Reply:
Gotcha, the reason why I'm pushing for bucket to be required is to support the following three cases:

1. Virtual Hosted Path: https://bucket-name.storage.googleapis.com
1. Path Style: https://storage.googleapis.com/bucket-name
1. Bucket Bound Hostname: http://cname.domain.tld

Based on discussions with the GCS team and research requiring the bucket name then storing it in policy is adaptable for these three cases where in contrast `key: bucket-name/object-name` will only work for Case 1&2, but not 3.